### PR TITLE
121 replace assert with check

### PR DESCRIPTION
**Pull request by milancurcic**
This PR implements the `check` subroutine, proposed to replace `assert` for internal testing. Specifically:

* Implements `check` as proposed [here](https://github.com/fortran-lang/stdlib/issues/121#issuecomment-602261670)
* Replaces existing calls to `assert` with calls to `check`
* Removes the definition of `assert` from `stdlib_experimental_error`

This PR doesn't cover making all `check` calls to print meaningful messages on failure, and to execute them with `warn=.true.`, to allow multiple test failures in a single run. These will be covered in a future PR.

This internal testing is meant as temporary until we can do it with `fpm test` as proposed in #162. For now, we consider this as a "good enough" solution that will help us move forward with development while [fpm](https://github.com/fortran-lang/fpm) is in the works.

Finally, there is also Brad's proposed implementation [here](https://github.com/fortran-lang/stdlib/issues/121#issuecomment-602632789) which we shouldn't ignore. If there is some support from the community for it, let's discuss and consider it.

### CODE REVIEW

File: src/stdlib_experimental_error.f90 [hunk 1]
```diff
@@ -1,34 +1,63 @@
...
+    !   * the message.
+    !
+    ! Arguments
+    ! ---------
+
+    logical, intent(in) :: condition
+    character(*), intent(in), optional :: msg
+    integer, intent(in), optional :: code
+    logical, intent(in), optional :: warn
+    character(*), parameter :: msg_default = 'Test failed.'
```
Comment (ivan-pi) [inline src/stdlib_experimental_error.f90]: Wouldn't it make more sense if the default message is `'Check failed.' just to be consistent with the subroutine name?

File: src/stdlib_experimental_error.f90 [hunk 2]
```diff
@@ -1,34 +1,63 @@
...
+
+    ! Examples
+    ! --------
+    !
+    ! ! Stops the program with exit code 1 and prints 'Test failed.'
+    ! call check(a == 5)
```
Comment (ivan-pi) [inline src/stdlib_experimental_error.f90]: There is an implicit assumption in the examples that `a` is not five. It remains to be discussed whether such examples in the docstrings should be self-contained.

**Comment by milancurcic** [comment]
@ivan-pi You're right. The latest commit fixes that.

**Comment by milancurcic** [comment]
Thanks, all. @everythingfunctional, let's do a separate issue and PR for your proposal when you're ready. Merging.

**Comment by jvdp1** [comment]
@milancurcic Do we want that users use both `check` and `error_stop` outside `stdlib`? OR are these only aimed for `stdlib`?
If users can use them outside `stdlib`, then I think it would be good to have some specs (like for other functions)  (I can give a start based on the comments in the code). Personally, I could use both outside `stdlib` ;)

**Comment by milancurcic** [comment]
@jvdp1 I think so. Sorry, I totally forgot about the spec. That should've been part of the original PR. Please go ahead with a draft spec if you don't mind. Thank you!

**Pull request merged**